bin/cron/hoc_student_name_cleanup

```diff
@@ -0,0 +1,196 @@
+last_row = HOC_ACTIVITY.order(:id).last
+if last_row.nil?
+  CDO.log.info "pegasus.hoc_activity table has no rows, exiting"
+  return
+end
+
+# Calculate the start of the retention window. Rows with a timestamp earlier
+# than this are candidates for cleanup.
+options[:retain_after_date] = Date.today - options[:retention_days]
+
+# If start_id was not specified, read the value from the db
+if options[:start_id].nil?
+  options[:start_id] = Properties.get(START_ID_PROPERTY_KEY) || 0
+end
+
+# If end_id was not specified, determine its value by looking for the first row
+# that is definitely inside the retention window. In the rare case where there
+# are no rows in the retention window, set end_id to 1 past the last row to
+# check all rows
+if options[:end_id].nil?
+  options[:end_id] = HOC_ACTIVITY.select(:id).
+    where {finished_at > options[:retain_after_date]}.
+    order(:finished_at, :id).
+    get(:id) || last_row[:id] + 1
+end
+
+# Options should not change from here on
+options.freeze
+CDO.log.info "Final calculated options: #{options}"
+
+# Main loop - read rows with names one batch at a time and null out entries
+# that are outside the retention window. It is important to keep the batch size
+# small because the name column is not indexed.
+batch_start_id = options[:start_id]
+first_retained_row = nil
+batch_count = 0
+while batch_start_id < options[:end_id]
+  ids_to_cleanup = []
+  last_row_id = nil
+
+  HOC_ACTIVITY.select(:id, :started_at, :finished_at, :pixel_started_at, :pixel_finished_at).
+    where(Sequel.lit('id >= ? AND id < ? AND name IS NOT NULL', batch_start_id, options[:end_id])).
+    order(:id).
+    first(options[:batch_limit]).
+    each do |row|
+      # A row needs to be cleaned up if it has a timestamp that is earlier than the
+      # retention window. Each row has four timestamps, we choose the first non-null
+      # value using the following order: finished_at, pixel_finished_at, started_at,
+      # pixel_started_at. In the rare case where all four timestamps are null, we
+      # assume that the row _should_ be cleaned up.
+      timestamp = row[:finished_at] || row[:pixel_finished_at] || row[:started_at] || row[:pixel_started_at]
+      if timestamp.nil? || timestamp < options[:retain_after_date]
+        ids_to_cleanup << row[:id]
+      else
+        # This is a row that should be retained. If it's the first such row, this
+        # is where we should start the next time we run this script
+        if first_retained_row.nil?
+          first_retained_row = row[:id]
+          CDO.log.info("First retained row: id=#{first_retained_row}")
+        end
+      end
+      last_row_id = row[:id]
+    end
+
+  # If no rows were returned, exit the main loop
+  break if last_row_id.nil?
+
+  # Wrap the update in a transaction so we can rollback the changes on a dry run
+  DB.transaction do
+    HOC_ACTIVITY.where(id: ids_to_cleanup).update(name: nil) unless ids_to_cleanup.empty?
```

Comment:
I'd like to monitor performance of this statement with you as it runs on a clone of the production database.